catkin_tools/execution/stages.py

```diff
@@ -0,0 +1,151 @@
+    """
+
+    def __init__(
+            self,
+            label,
+            cmd,
+            cwd=None,
+            env=None,
+            env_overrides=None,
+            shell=False,
+            emulate_tty=True,
+            stderr_to_stdout=False,
+            occupy_job=True,
+            logger_factory=IOBufferProtocol.factory):
+        """
+        :param label: The label for the stage
+        :param command: A list of strings composing a system command
+        :param protocol: A protocol class to use for this stage
+
+        :parma cmd: The command to run
+        :param cwd: The directory in which to run the command (default: os.getcwd())
+        :param env: The base environment. (default: {})
+        :param env_overrides: The variables that override the base environment. (default: {})
+        :param shell: Whether to run the command in "shell" mode. (default: False)
+        :param emulate_tty: Support colored output (default: True)
+        :param stderr_to_stdout: Pipe stderr to stdout (default: False)
+
+        :param occupy_job: Whether this stage should wait for a worker from the job server (default: True)
+        :param logger_factory: The factory to use to construct a logger (default: IOBufferProtocol.factory)
+        """
+
+        if not type(cmd) in [list, tuple] or not all([isinstance(s, basestring) for s in cmd]):
+            raise ValueError('Command stage must be a list of strings: {}'.format(cmd))
+        super(CommandStage, self).__init__(label, logger_factory, occupy_job)
+
+        # Store environment overrides
+        self.env_overrides = env_overrides or {}
+
+        # Override base environment
+        env = env or {}
+        env.update(self.env_overrides)
+
+        self.async_execute_process_kwargs = {
+            'cmd': cmd,
+            'cwd': cwd or os.getcwd(),
+            'env': env,
+            'shell': shell,
+            # Emulate tty for cli colors
+            'emulate_tty': emulate_tty,
+            # Capture stderr and stdout separately
+            'stderr_to_stdout': stderr_to_stdout,
+        }
+
+    def update_env(self, base_env):
+        """Update the environment for this stage with the env args passed to the constructor."""
+        self.async_execute_process_kwargs['env'].update(base_env)
+        self.async_execute_process_kwargs['env'].update(self.env_overrides)
+
+    def get_repro(self, verb, jid):
+        """Get a command line to reproduce this stage."""
+
+        # Define the base env command
+        env_cmd = 'catkin {} --env {}'.format(verb, jid)
+
+        # Add additional env args
+        env_override_repro = ' '.join([
+            '{}={}'.format(k, cmd_quote(v))
+            for k, v in self.env_overrides.items()
+        ])
+        if len(env_override_repro) > 0:
+            env_cmd = 'echo "$({}) {}"'.format(env_cmd, env_override_repro)
+
+        # Return the full command
+        return 'cd {}; {} | xargs -I %ENV% env %ENV% {}; cd -'.format(
```

Comment:
Let's not use `xargs` here. For a few reasons:

- It's confusing and not obvious to new users.
- It doesn't work on Windows. (Important long term for ROS 2 support)
- It doesn't work the same on all OS's, e.g. BSD and OS X's `xargs` has different options than GNU version.

On my machine, when trying to reproduce a build using this:

```
$ cd /Users/william/indigo/build/rqt_nav_view; catkin build --env rqt_nav_view | xargs -I %ENV% env %ENV% /usr/local/bin/cmake /Users/william/indigo/src/rqt_robot_plugins/rqt_nav_view --no-warn-unused-cli -DCATKIN_DEVEL_PREFIX=/Users/william/indigo/devel -DCMAKE_INSTALL_PREFIX=/Users/william/indigo/install; cd -
env: %ENV%: No such file or directory
~/indigo
```

My guess is that this is due to a difference in the `xargs` on OS X versus the one on Linux.

I'd propose a new option for `build` or a new verb to handle this case:

```
$ catkin build --run-with-env-for-package <package name> <command> [arg1 [argN] [--]]
```

Or something like this:

```
$ catkin run_with_env_for_package --package <package name> <command> [arg1 [argN] [--]]
```

The description of this commands can explain that it works like `xargs -I {} env {} ...` but is portable.

The least effort change would be to figure out how to change the output so it would work on OS X and Linux.